frontend/src/lib/k8s/api/v2/webSocket.ts

```diff
@@ -167,3 +169,93 @@ export function useWebSocket<T>({
     };
   }, [enabled, url, protocols, type, cluster]);
 }
+
+export type WebSocketConnectionRequest<T> = {
+  cluster: string;
+  url: string;
+  onMessage: (data: T) => void;
+};
+
+/**
+ * Creates or joins mutiple existing WebSocket connections
+ *
+ * @param url - endpoint URL
+ * @param options - WebSocket options
+ */
+export function useWebSockets<T>({
+  connections,
+  enabled = true,
+  protocols,
+  type = 'json',
+}: {
+  enabled?: boolean;
+  connections: Array<WebSocketConnectionRequest<T>>;
+  /**
+   * Any additional protocols to include in WebSocket connection
+   */
+  protocols?: string | string[];
+  /**
+   * Type of websocket data
+   */
+  type?: 'json' | 'binary';
+}) {
+  useEffect(() => {
+    if (!enabled) return;
+
+    let isCurrent = true;
+
+    function connect({ cluster, url, onMessage }: WebSocketConnectionRequest<T>) {
+      const connectionKey = cluster + url;
+
+      if (sockets.has(connectionKey)) return () => {};
+
+      // Add new listener for this URL
+      listeners.set(connectionKey, [...(listeners.get(connectionKey) ?? []), onMessage]);
+
+      // Mark socket as pending, so we don't open more than one
+      sockets.set(connectionKey, 'pending');
+
+      let ws: WebSocket | undefined;
+      openWebSocket(url, { protocols, type, cluster, onMessage }).then(socket => {
+        ws = socket;
+
+        // Hook was unmounted while it was connecting to WebSocket
+        // so we close the socket and clean up
+        if (!isCurrent) {
+          ws.close();
+          sockets.delete(connectionKey);
+          return;
+        }
+
+        sockets.set(connectionKey, ws);
+      });
+
+      return () => {
+        const connectionKey = cluster + url;
+
+        // Clean up the listener
+        const newListeners = listeners.get(connectionKey)?.filter(it => it !== onMessage) ?? [];
+        listeners.set(connectionKey, newListeners);
+
+        // No one is listening to the connection
+        // so we can close it
+        if (newListeners.length === 0) {
+          const maybeExisting = sockets.get(connectionKey);
+          if (maybeExisting) {
+            if (maybeExisting !== 'pending') {
+              maybeExisting.close();
+            }
+            sockets.delete(connectionKey);
+          }
+        }
+      };
+    }
+
+    const disconnectCallbacks = connections.map(endpoint => connect(endpoint));
+
+    return () => {
+      isCurrent = false;
+      disconnectCallbacks.forEach(fn => fn());
+    };
+  }, [enabled, type, connections, protocols]);
```

Comment:
Maybe add a note about the dependencies here?
`// @todo: We maybe have to be careful with the arrays here. The only caller of this function so far creates a new array every time (cached with useMemo) but other callers may not.`

Reply:
it's not really a todo, users of this functions need to make sure to keep references stable
added a comment to properties to make it clear that they should be stable